File: lessons/46-code-execution-and-files-api/python/legacy.py
"""Lesson 46 - the Academy's tool version and result parsing, preserved.

LEGACY. Two separate problems:

  1. `code_execution_20250522` is the legacy Python-only version.
  2. The result parsing looks for a bare `code_execution_output` block, which
     current versions do not return. That check silently matches nothing - the
     code does not error, it just finds no output.

This file runs the legacy tool version if the API still accepts it, and shows
BOTH parsers side by side against the same response.

    uv run lesson 46 legacy
"""

import anthropic
from course.config import MODEL

client = anthropic.Anthropic()

LEGACY_TOOL = {"type": "code_execution_20250522", "name": "code_execution"}

print(f"Requesting {LEGACY_TOOL['type']} on {MODEL}...\n")

try:
    response = client.messages.create(
        model=MODEL,
        max_tokens=3000,
        tools=[LEGACY_TOOL],
        messages=[
            {
                "role": "user",
                "content": "Compute the mean and standard deviation of 1..10.",
            }
        ],
    )
except anthropic.BadRequestError as error:
    print(f"400 from the API: {error.message}")
    print("\nUse code_execution_20260521 - see example.py.")
    raise SystemExit(0) from None

print("blocks:", [block.type for block in response.content])
print()

# LEGACY parser: looks for a block type current versions do not emit.
legacy_hits = [b for b in response.content if b.type == "code_execution_output"]
print(f"legacy parser  (code_execution_output):        {len(legacy_hits)} block(s)")

# CURRENT parser.
current_hits = [
    b for b in response.content if b.type == "bash_code_execution_tool_result"
]
print(f"current parser (bash_code_execution_tool_result): {len(current_hits)} block(s)")

for block in current_hits:
    result = block.content
    if result.type == "bash_code_execution_result":
        print(f"\nstdout:\n{result.stdout.rstrip()}")

print(
    "\nThe legacy parser does not raise. It finds nothing, and a program built\n"
    "on it silently reports no output. That is the failure mode to watch for."
)

File: lessons/46-code-execution-and-files-api/typescript/legacy.ts
/**
 * Lesson 46 - the Academy's tool version and result parsing, preserved.
 *
 * LEGACY. Two separate problems:
 *
 *   1. `code_execution_20250522` is the legacy Python-only version.
 *   2. The result parsing looks for a bare `code_execution_output` block, which
 *      current versions do not return. That check silently matches nothing -
 *      the code does not error, it just finds no output.
 *
 * This runs the legacy tool version if the API still accepts it, and shows BOTH
 * parsers side by side against the same response.
 *
 *   npm run lesson -- 46 legacy
 */
import Anthropic from "@anthropic-ai/sdk";
import { MODEL } from "../../../shared/typescript/config.ts";

const client = new Anthropic();

const LEGACY_TOOL = {
  type: "code_execution_20250522",
  name: "code_execution",
} as unknown as Anthropic.Messages.ToolUnion;

console.log(`Requesting code_execution_20250522 on ${MODEL}...\n`);

let response: Anthropic.Message;
try {
  response = await client.messages.create({
    model: MODEL,
    max_tokens: 3000,
    tools: [LEGACY_TOOL],
    messages: [
      { role: "user", content: "Compute the mean and standard deviation of 1..10." },
    ],
  });
} catch (error) {
  if (error instanceof Anthropic.BadRequestError) {
    console.log(`400 from the API: ${error.message}`);
    console.log("\nUse code_execution_20260521 - see example.ts.");
    process.exit(0);
  }
  throw error;
}

console.log("blocks:", response.content.map((block) => block.type).join(", "));
console.log();

// LEGACY parser: looks for a block type current versions do not emit. It is not
// in the ContentBlock union any more, hence the cast.
const legacyHits = response.content.filter(
  (block) => (block.type as string) === "code_execution_output",
);
console.log(
  `legacy parser  (code_execution_output):          ${legacyHits.length} block(s)`,
);

// CURRENT parser.
const currentHits = response.content.filter(
  (block) => block.type === "bash_code_execution_tool_result",
);
console.log(
  `current parser (bash_code_execution_tool_result): ${currentHits.length} block(s)`,
);

for (const block of currentHits) {
  if (block.content.type === "bash_code_execution_result") {
    console.log(`\nstdout:\n${block.content.stdout.trimEnd()}`);
  }
}

console.log(
  "\nThe legacy parser does not throw. It finds nothing, and a program built\n" +
    "on it silently reports no output. That is the failure mode to watch for.",
);
